content/blog/pitfalls-of-safe-rust/index.md

```diff
@@ -0,0 +1,714 @@
+
+
+## Protect Against Time-of-Check to Time-of-Use (TOCTOU)
+
+This is a more advanced topic, but it's important to be aware of it.
+TOCTOU is a class of software bugs caused by attackers exploiting the gap between the time a condition is checked and the time it is used. 
+Simply put, an attacker could rug-pull the value from under your feet
+and you wouldn't notice because it happens exactly between your check and use.
+
+```rust
+// DON'T: Check and use in separate operations
+fn process_file(path: &Path) -> Result<(), io::Error> {
+    if path.exists() {  // Check
+        let content = fs::read(path)?;  // Use - file might have changed!
+        // Process content
+    }
+    Ok(())
+}
+```
+
+Rust supports atomic file operations to prevent this kind of attack.
+
+```rust
+// DO: Use atomic operations when possible
+fn process_file(path: &Path) -> Result<(), io::Error> {
+    let file = fs::OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create_new(true)  // Atomic operation
+        .open(path)?;
+    // Process file
+    Ok(())
+}
+```
+
+It's a bit more verbose, but it's safer.
+I think more Rustaceans should know about this pattern.
+
+
+## Use Constant-Time Comparison for Sensitive Data
+
+Timing attacks are a nifty way to extract information from your application.
+The idea is that the time it takes to compare two values can leak information about them.
+For example, the time it takes to compare two strings can reveal how many characters are correct.
+Therefore, for production code, be careful with regular equality checks when handling sensitive data like passwords.
+
+```rust
+// DON'T: Use regular equality for sensitive comparisons
+fn verify_password(stored: &[u8], provided: &[u8]) -> bool {
+    stored == provided  // Vulnerable to timing attacks!
+}
+
+// DO: Use constant-time comparison
+use subtle::{ConstantTimeEq, Choice};
+
+fn verify_password(stored: &[u8], provided: &[u8]) -> bool {
+    stored.ct_eq(provided).unwrap_u8() == 1
+}
+```
+
+## Don't Accept Unbounded Input
+
+Protect Against Denial-of-Service Attacks with Resource Limits.
+These happen when you accept unbounded input, e.g. a huge request body
+which might not fit into memory.
+
+```rust
+// DON'T: Accept unbounded input
+fn process_request(data: &[u8]) -> Result<(), Error> {
+    let decoded = decode_data(data)?;  // Could be enormous!
+    // Process decoded data
+    Ok(())
+}
+```
+
+Instead, set explicit limits for your accepted payloads:
+
+```rust
+const MAX_REQUEST_SIZE: usize = 1024 * 1024;  // 1MB
+
+fn process_request(data: &[u8]) -> Result<(), Error> {
+    if data.len() > MAX_REQUEST_SIZE {
+        return Err(Error::RequestTooLarge);
+    }
+    
+    let decoded = decode_data(data)?;
+    
+    // For added safety, you can also check the decoded size 
+    if decoded.len() > MAX_REQUEST_SIZE * 2 {  
+        return Err(Error::DecodedDataTooLarge);
+    }
```

Comment:
That's not added safety, that's highly confusing and bad advice. What does the decode do? Does it strictly shorten the data? Can it increase the size? Or is it simply a 1:1 translation?

Just leave it out, but at the very least use a different constant.

Reply:
Removed that part, thanks.